# Pandas Library 
Pandas is the most popular library of python used for data analysis.
To work with it, first import the library.(you can use alias for simplicity)

In [12]:
import pandas as pd

Two main concepts in pandas are:

1. DataFrame
2. Series

We will create our own data using these as follows

**DataFrame** <br>
DataFrame is a table. It contains an array of entries. Each entry corresponds to row and coloumn.
For example: 31 is a entry whose row=0, column = yes

In [13]:
data = pd.DataFrame({'yes':[31,3],'no':[33,9]})
print(data)

   yes  no
0   31  33
1    3   9


As you can see, the table is automatically given index values which starts for 0.
If you wish to use give specific values to index, do  follows:

In [14]:
data = pd.DataFrame({'Bob':["I likes it","Bland"],
                     'Alice':["It was awful","I liked it"]},
                     index = ['Product A','Product B'])
print(data)

                  Bob         Alice
Product A  I likes it  It was awful
Product B       Bland    I liked it


**Series**:
Series is a sequence of data values. If DataFrame is Table, series is a List.

In [15]:
data = pd.Series([1,2,3,4])
data

0    1
1    2
2    3
3    4
dtype: int64

Series is just a colum of DataFrame. So, you can assign a row lable using `index` parameter. However, a Series does not have a column name. It only has one name overall.

In [16]:
data = pd.Series([30,40,50],index=['Subject A','Subject B','Subject c'],name = 'Marks of a Student')
data

Subject A    30
Subject B    40
Subject c    50
Name: Marks of a Student, dtype: int64

### Reading data files 

As of now, we have seen how to create our own data. But, most of the time, we won't be creating data by hand. Instead, we will be working with the already existing data.
Data can be stored in any format.By far the most used is CSV files. <br>
CSV stands for Comma separeted values.
When you open a CSV file, you will get to see the data values separated by commas and hence the name CSV. <br>
To read the data into DataFrame, we use `pd.read_csv()` function 

In [17]:
data = pd.read_csv("/kaggle/input/datasets/spoortibadiger/network-intrusion-dataset/Wednesday-workingHours.pcap_ISCX.csv")

Before we move further, let us know how to upload the datasets to kaggle and sue them:
1. **Add the dataset (if not already done)** <br>
In your notebook, look at the right-hand sidebar (in the Notebook editor). Click "Add Input" (sometimes shown as a "+" next to "Input"). Search for your dataset by name and click the + to add it. This mounts it — it doesn't copy files into your working code.
2. **Find the exact path** <br>
Once added, Kaggle places datasets under:<br>
`/kaggle/input/<dataset-folder-name>/`
3. To see the exact folder and file names (they don't always match what you expect), run this in a code cell: <br>
```
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))
```
Now use the path you get after running this code in the `read_csv()`

We can use the `shape` attribute to check how large the resulting DataFrame is:

In [18]:
data.shape

(692703, 79)

The above output tells us that the new DataFrame created have total of 692703 records split across 79 columns.
<br>
To examine the content of the dataset, we can use `head()` command which gives us the first 5 rows of the dataset.

In [19]:
data.head()

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
0,80,38308,1,1,6,6,6,6,6.000000,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
1,389,479,11,5,172,326,79,0,15.636364,31.449238,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
2,88,1095,10,6,3150,3150,1575,0,315.000000,632.561635,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
3,389,15206,17,12,3452,6660,1313,0,203.058823,425.778474,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
4,88,1092,9,6,3150,3152,1575,0,350.000000,694.509719,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


If you wish to fetch more than just 5 rows then specify the nymber of rows in the `head()` command.

In [20]:
data.head(10)

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
0,80,38308,1,1,6,6,6,6,6.000000,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
1,389,479,11,5,172,326,79,0,15.636364,31.449238,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
2,88,1095,10,6,3150,3150,1575,0,315.000000,632.561635,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
3,389,15206,17,12,3452,6660,1313,0,203.058823,425.778474,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
4,88,1092,9,6,3150,3152,1575,0,350.000000,694.509719,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
5,389,433,11,4,172,326,79,0,15.636364,31.449238,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
6,88,1088,9,6,3150,3152,1575,0,350.000000,694.509719,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
7,80,579225,132,150,160,320799,160,0,1.212121,13.926212,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
8,49666,3,2,0,12,0,6,6,6.000000,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
9,49413,4,3,0,18,0,6,6,6.000000,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


If you wish to checkout the data from last, then use `tail()` command. This command gives the last 5 rows of the dataset.

In [21]:
data.tail()

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
692698,53,32215,4,2,112,152,28,28,28.0,0.00000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692699,53,324,2,2,84,362,42,42,42.0,0.00000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692700,58030,82,2,1,31,6,31,0,15.5,21.92031,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692701,53,1048635,6,2,192,256,32,32,32.0,0.00000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692702,53,94939,4,2,188,226,47,47,47.0,0.00000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


In [23]:
#to fetch the last 10 rows of the dataset
data.tail(10)

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
692693,53,158,2,2,82,114,41,41,41.0,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692694,35450,33,1,1,0,0,0,0,0.0,0.000000,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692695,53,46982,2,2,58,446,29,29,29.0,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692696,80,11512230,5,5,672,3525,672,0,134.4,300.527536,...,32,821.0,0.0,821,821,6506213.0,0.0,6506213,6506213,DoS GoldenEye
692697,49346,145,1,1,0,0,0,0,0.0,0.000000,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692698,53,32215,4,2,112,152,28,28,28.0,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692699,53,324,2,2,84,362,42,42,42.0,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692700,58030,82,2,1,31,6,31,0,15.5,21.920310,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692701,53,1048635,6,2,192,256,32,32,32.0,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
692702,53,94939,4,2,188,226,47,47,47.0,0.000000,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


The `pd.read_csv()` has over 30 optional parameters. For example, you can see in this dataset taht the CSV file has a built-in index, which pandas did not pick up automatically. To make pandas use that column for the index (instead of creating a new one from scratch), we can specify an `index_col`. 

In [25]:
data = pd.read_csv("/kaggle/input/datasets/spoortibadiger/network-intrusion-dataset/Wednesday-workingHours.pcap_ISCX.csv",index_col=0)
data.head()

,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,Bwd Packet Length Max,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
Destination Port,,,,,,,,,,,,,,,,,,,,,
80,38308,1,1,6,6,6,6,6.000000,0.000000,6,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
389,479,11,5,172,326,79,0,15.636364,31.449238,163,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
88,1095,10,6,3150,3150,1575,0,315.000000,632.561635,1575,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
389,15206,17,12,3452,6660,1313,0,203.058823,425.778474,3069,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
88,1092,9,6,3150,3152,1575,0,350.000000,694.509719,1576,...,32,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


Now, let us learn how to save a DataFarme as CSV to disk using `to_csv`.
Let us create a DataFrame called animals and save it as a CSV file to disk under the name `cows_and_goats.csv`.

In [26]:
animals = pd.DataFrame({'Cows': [12, 20], 'Goats': [22, 19]}, index=['Year 1', 'Year 2'])
animals

,Cows,Goats
Year 1,12,22
Year 2,20,19


In [28]:
#use the function to_csv() to convert the datafame to csv
animals.to_csv("cows_and_goats.csv")

### Indexing, Selecting and assigning